src/index.js

```diff
@@ -0,0 +1,108 @@
+// @flow
+import type { RemoteUser, Callback } from '@verdaccio/types';
+import type { verdaccioConfigAuthMemory, VerdaccioConfigApp } from '../types';
+/**
+ * Verdaccio auth memory
+ */
+export default class Memory {
+  // flow types
+  config: Object;
+  verdaccioConfig: Object;
+  users: Object;
+  max_users: number;
+  constructor(
+    config: verdaccioConfigAuthMemory,
+    verdaccioConfig: VerdaccioConfigApp
+  ) {
+    this.config = config;
+    this.verdaccioConfig = verdaccioConfig;
+    this.users = config.users || {};
+    this.max_users = verdaccioConfig.max_users || Infinity;
+  }
+
+  authenticate(user: string, password: string, done: Callback): Callback {
+    const { users } = this;
+    const userCredentials = users[user];
+
+    if (!userCredentials) {
+      return done(null, false);
+    }
+
+    if (password !== userCredentials.password) {
+      const error = Error('Unauthorized access');
+      // $FlowFixMe
+      error.status = 401;
+      return done(error);
+    }
+
+    // authentication succeeded!
+    // return all usergroups this user has access to;
+    return done(null, [user]);
+  }
+
+  adduser(user: Object, password: string, done: Callback): Callback {
+    const { users, max_users } = this;
+    if (users[user]) {
+      return done(null, true);
+    }
+
+    if (max_users && Object.keys(users).length >= max_users) {
+      let error = Error('maximum amount of users reached');
+      // $FlowFixMe
+      error.status = 409;
+      return done(error);
+    }
+
+    this.users[user] = { name: user, password };
+
+    return done(null, user);
+  }
+
+  allow_access(user: RemoteUser, pkg: Object, done: Callback): Callback {
+    if (pkg.access.includes('$all') || pkg.access.includes('$anonymous')) {
+      return done(null, true);
+    }
+
+    const { name } = user;
+
+    if (!name) {
+      const error = Error('not allowed to access package');
```

Comment:
For errors we should not use native `Error`, instead in verdaccio we use https://github.com/verdaccio/local-storage/blob/master/src/local-fs.js#L8 

For now we can let it like that, but, in some point we have to centralize this https://github.com/verdaccio/verdaccio/blob/master/src/lib/utils.js#L354 in a package